### PR TITLE
[Nemo-UX] Add fabric-API for manual forward-pass

### PR DESCRIPTION
**Pull request by marcromeyn**
# What does this PR do ?

Adds the Fabric-API to nemo 2

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
from nemo import lightning as nl
from nemo.collections.llm import MistralModel, MockDataModule


trainer = nl.Trainer(
    devices=2, 
    strategy=nl.MegatronStrategy(tensor_model_parallel_size=2),
    plugins=nl.MegatronMixedPrecision(precision='16-mixed')
)

fabric = trainer.to_fabric()

# Import model
model = fabric.import_model("hf://mistralai/Mistral-7B-v0.1", MistralModel)

# Load model
# model = fabric.load_model("/workspaces/llm-workspace/nemo_experiments/default/2024-07-02_01-14-20/checkpoints/default--reduced_train_loss=nan-epoch=0")


fabric.print(type(model))
# The model here will be the MistralModel wrapped by the MegatronStrategy

# Some dummy data
dataset = MockDataModule(tokenizer=model.module.tokenizer).train_dataloader()
batch = next(iter(dataset))
fabric.print(batch)

# Manual model forward
output = model(
    batch,
    forward_step=model.config.forward_step_fn,
    loss_reduction=model.module.validation_loss_reduction(),
)
fabric.print(output)

```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by akoumpa — APPROVED** [review]
LGTM, thanks.

**Review by akoumpa — APPROVED** [review]
